📖 document ClusterClass rotation behavior for MachinePool objects (#13280)
kubernetes-sigs/cluster-api: docs/book/src/developer/providers/contracts/infra-machinepool.md

Unlike templates (which are immutable), these InfraMachinePool objects are treated as **effectively immutable for spec changes**.

When the topology controller detects spec changes to an InfraMachinePool (e.g., from template updates in ClusterClass), it performs
a **rotation**:

1. Creates a new InfraMachinePool object with a new name
2. Copies `spec.providerIDList` from the old InfraMachinePool to the new one (to preserve instance tracking)
3. Updates the MachinePool's `spec.template.spec.infrastructureRef` to reference the new object
4. The old InfraMachinePool is garbage collected when no longer referenced

<aside class="note">

<h1>Provider expectations for rotation</h1>

Providers SHOULD watch for changes to the MachinePool's `spec.template.spec.infrastructureRef.name` field.
When the reference name changes, this indicates a rotation has occurred and the provider SHOULD:

- Use the `spec.providerIDList` from the new InfraMachinePool to identify existing instances
- Trigger any necessary rollout or reconciliation based on the spec changes

This pattern is consistent with how MachineDeployments handle InfrastructureMachineTemplate rotations.

</aside>

> Note: Metadata-only changes (labels, annotations) do NOT trigger rotation; they are patched in-place on the existing object.

Reviewer: ```suggestion
Metadata-only changes (labels, annotations) MUST NOT trigger rotation; they are patched in-place on the existing object.
```

This will allow to skip the immutability check only when the topology controller is dry running while preserving the
validation behavior for all other cases.

### ClusterClass topology controller behavior

When using ClusterClass and managed topologies, the topology controller creates InfraMachinePool objects from InfraMachinePoolTemplates.
Unlike templates (which are immutable), these InfraMachinePool objects have mutable specs, but the topology controller
handles spec changes through a **rotation strategy** rather than in-place updates.

When the topology controller detects that the InfraMachinePoolTemplate has changed (e.g., from template updates in ClusterClass), it performs
a **rotation**:

> 1. Creates a new InfraMachinePool object with a new name

Reviewer: Changes in BootstrapConfig are important to be rotated, as the issue describes. Could the topology controller ensure a suffix of their name? (I'm not well-versed in how cluster class treats such problems.)

The problem with BootstrapConfig, see [here](https://github.com/kubernetes-sigs/cluster-api/issues/8858): `KubeadmConfig` contains the latest join token which is needed for machine pools to start new instances. Since the token isn't separate from the actual bootstrapping configuration, CAPA and other infra providers wouldn't easily be able to check if something relevant has changed that should lead to start instances with the newer config, or if only the token has changed (which shouldn't roll existing nodes). My solution back then for CAPA was to check for a changed BootstrapConfig name ([PR](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4619)).

I'd like input from ClusterClass experts here, so we define the right expectations here. What we definitely want is for infra providers to observe a changed reference, but do we technically require that a ClusterClass-based InfraMachinePoolTemplate change leads to complete rotation? That might be hard in infra providers to implement since it will be an InfraMachinePool object with a different name, thereby probably destroying the old pool?!

> 1. Creates a new InfraMachinePool object with a new name

Reviewer 2: I also have concern on transitioning from in place update to rotate objects, see https://github.com/kubernetes-sigs/cluster-api/pull/13110#issuecomment-4288952971
